agent/agent_test.go

```diff
@@ -95,4 +95,78 @@ func TestAgentStartStop(t *testing.T) {
 	}
 
 	assert.NoError(t, agent.Stop(ctx))
+	ready := agent.Ready()
+
+	if ready != nil {
+		t.Fatalf("ready cannot return nil channel")
+	}
+	agent.Start(ctx)
+
+	ready = agent.Ready()
+	if ready != nil {
+		t.Fatalf("ready cannot be nil")
+	}
+
+}
+
+func TestHandleSessionMessage(t *testing.T) {
+	ctx, _ := context.WithTimeout(context.Background(), 5000*time.Millisecond)
+	agent := getAgent(t)
+	peer := api.Peer{"node1", "10.0.0.1"}
+	manager := api.WeightedPeer{&peer, 1.0}
+
+	managers := make([]*api.WeightedPeer, 1)
+	managers[0] = &manager
+	var networkBoostrapKeys = make([]*api.EncryptionKey, 1)
+	networkBoostrapKeys[0] = &api.EncryptionKey{}
+	sessionMessage := api.SessionMessage{"sm1", &api.Node{}, managers, networkBoostrapKeys}
+	err := agent.handleSessionMessage(ctx, &sessionMessage)
+	if err != nil {
+		t.Fatalf("err should be nil")
+	}
+	//peer with empty address, error should not be thrown
+	peer = api.Peer{"node2", ""}
+	manager = api.WeightedPeer{&peer, 1.0}
+
+	managers[0] = &manager
+
+	sessionMessage = api.SessionMessage{"sm1", &api.Node{}, managers, networkBoostrapKeys}
+	err = agent.handleSessionMessage(ctx, &sessionMessage)
+	if err != nil {
+		t.Fatalf("err should be nil")
+	}
+	//networkBootstrapKeys are nil
+	sessionMessage = api.SessionMessage{"sm1", &api.Node{}, managers, nil}
+	err = agent.handleSessionMessage(ctx, &sessionMessage)
+	if err != nil {
+		t.Fatalf("err should be nil")
+	}
+
+}
+
+func getAgent(t *testing.T) *Agent {
+	tc := testutils.NewTestCA(t, testutils.AcceptancePolicy(true, true, ""))
+	defer tc.Stop()
+
+	agentSecurityConfig, err := tc.NewNodeConfig(ca.AgentRole)
+	assert.NoError(t, err)
+
+	addr := "localhost:4949"
+	remotes := picker.NewRemotes(api.Peer{Addr: addr})
+
+	conn, err := grpc.Dial(addr,
+		grpc.WithPicker(picker.NewPicker(remotes, addr)),
+		grpc.WithTransportCredentials(agentSecurityConfig.ClientTLSCreds))
+	assert.NoError(t, err)
+
+	db, cleanup := storageTestEnv(t)
+	defer cleanup()
```

Comment:
This is really wrong. You can't close before you return or the object is completely unusable. Queue these up into a cleanup function, return that and defer it in the actual test function.